src/Hpack/Exception.hs

```diff
@@ -0,0 +1,72 @@
+{-# LANGUAGE RecordWildCards #-}
+
+-- | Exceptions thrown by the Hpack library.
+module Hpack.Exception
+  ( HpackException (..)
+    -- * Re-export of types used in Hpack exceptions
+  , FilePath
+  , ParseException (..)
+  , ProgramName (..)
+  , Status (..)
+  , Version (..)
+  , URL
+  , YamlException (..)
+  , YamlMark (..)
+  ) where
+
+import           Control.Exception (Exception)
+import qualified Data.ByteString.Char8 as B
+import           Data.List (intercalate)
+import           Data.Typeable (Typeable)
+import           Data.Version (Version (..), showVersion)
+import           Data.Yaml
+                   (ParseException (..), YamlException (..), YamlMark (..))
+import           Network.HTTP.Types.Status (Status (..))
+
+import           Hpack.ProgramName (ProgramName (..))
+
+-- | Type synonyn representing URLs.
+type URL = String
+
+-- | Type representing exceptions thrown by functions exported by the modules of
+-- the Hpack library.
+data HpackException
+  = CycleInDefaultsException ![FilePath]
+  | DecodeValueException !FilePath !String
+  | DefaultsFileNotFound !FilePath
+  | DefaultsFileUrlNotFound !URL
+  | DownloadingFileFailed !URL !Status
+  | HpackParseException !FilePath !ParseException
+  | HpackVersionUnsupported !ProgramName !FilePath !Version !Version
+  deriving Typeable
+
+instance Show HpackException where
+  show (CycleInDefaultsException files) =
+    "cycle in defaults (" ++ intercalate " -> " files ++ ")"
+  show (DecodeValueException file s) =
+    file ++ ": " ++ s
+  show (DefaultsFileNotFound file) =
+    "Invalid value for \"defaults\"! File " ++ file ++ " does not exist!"
+  show (DefaultsFileUrlNotFound url) =
+    "Invalid value for \"defaults\"! File " ++ url ++ " does not exist!"
+  show (DownloadingFileFailed url status) =
+    "Error while downloading " ++ url ++ " (" ++ formatStatus status ++ ")"
+   where
+    formatStatus :: Status -> String
+    formatStatus (Status code message) = show code ++ " " ++ B.unpack message
+  show (HpackParseException file e) = file ++ case e of
+    AesonException s -> ": " ++ s
+    InvalidYaml (Just (YamlException s)) -> ": " ++ s
+    InvalidYaml (Just (YamlParseException{..})) ->
+      let YamlMark{..} = yamlProblemMark
+      in  ":" ++ show yamlLine ++ ":" ++ show yamlColumn ++ ": " ++
+            yamlProblem ++ " " ++ yamlContext
+    err -> ": " ++ show err
+  show (HpackVersionUnsupported (ProgramName progName) file wanted supported) =
+    "The file " ++ file ++ " requires version " ++ showVersion wanted ++
+    " of the Hpack package specification, however this version of " ++
+    progName ++ " only supports versions up to " ++ showVersion supported ++
+    ". Upgrading to the latest version of " ++ progName ++
+    " may resolve this issue."
```

Comment:
@mpilgrem as a sidenote, for exceptions you want to derive the `Show` instance and define `displayException` instead. The unwritten law of `Show` is that it should produce syntactically valid Haskell code when possible.

Reply:
Thanks for that advice. I had not appreciated that, and simply copied the pattern used in Stack and Pantry. I have now changed to `displayException` here.